Refs #30593 -- Fixed introspection of check constraints columns on MariaDB.

hramezani opened the pull request:
Continue to https://github.com/django/django/pull/11566.

felixxm commented on django/db/backends/mysql/introspection.py:
```diff
@@ -202,13 +203,15 @@ def get_constraints(self, cursor, table_name):
+                for token in tokens:
+                    if token.ttype in (sqlparse.tokens.Name, sqlparse.tokens.Literal.String.Single):
+                        if token.value[1:-1] in columns:
+                            cols.add(token.value[1:-1])
```
Can you explain why we don't use the same logic as in [the SQLite introspection](https://github.com/django/django/blame/c7bef16a749ccbe2dbc8ce6b697dc80146584384/django/db/backends/sqlite3/introspection.py#L286-L291), i.e.
- we use `sqlparse.tokens.Literal.String.Single` instead of `sqlparse.tokens.Literal.String.Symbol`,
- we don't use `sqlparse.tokens.Keyword`, and
- we unquote `sqlparse.tokens.Name`.

:thinking: 

I think we should add MySQL/MariaDB-specific tests of check constraints introspection (see [`test_check_constraint`](https://github.com/django/django/commit/782d85b6dfa191e67c0f1d572641d8236c79174c#diff-4379f72f6de99e27a38abec7669669a3R105-R117)).

hramezani replied:
> Can you explain why we don't use the same logic as in [the SQLite introspection](https://github.com/django/django/blame/c7bef16a749ccbe2dbc8ce6b697dc80146584384/django/db/backends/sqlite3/introspection.py#L286-L291), i.e.

If we use the same logic as `SQLite introspection`, [introspection.tests.IntrospectionTests.test_get_constraints](https://github.com/django/django/blob/c7bef16a749ccbe2dbc8ce6b697dc80146584384/tests/introspection/tests.py#L215) will fail because when we run the test ``` `up_votes` >= 0 ``` will added to `information_schema.check_constraints` table.
When we parse this entry we have these 3 tokens:
- ``` `up_votes` ``` with type `Token.Name`
- `>=` with type `Token.Operator.Comparison`
- `0` with type `Token.Literal.Number.Integer`

So, ``` `up_votes` ``` will fall into the first block:
```
    if token.ttype in (sqlparse.tokens.Name, sqlparse.tokens.Keyword):
        if token.value in columns:
            cols.append(token.value)
```
and in this block we check `if token.value in columns`, and `token.value` is ``` `up_votes` ```.

I think it would be good to replace `token.value[1:-1]` with ```token.value.strip('`')``` to handle the check constraint like `JSON_VALID`.

About the token types, I just used these types (`tokens.Name`, `tokens.Literal.String.Single`)
because they are matched with our test cases.

felixxm replied:
It's hard to say without tests. IMO we should add `_parse_constraint_columns(sql)` hook and various tests in the `tests/backends/mysql/test_introspection.py`.

hramezani replied:
I will do it.

hramezani commented:
@felixxm I added the `_parse_constraint_columns` function and also some tests in `tests/backends/mysql/test_introspection.py`.
I created the example table in [mariadb check constraint](https://mariadb.com/kb/en/library/constraint/#check-constraints) and added the `check_clause` from  `information_schema.check_constraints` table to tests. if you have something else please let me know.

Also, I just check tokens with type `tokens.Name` because here we just want to extract columns from check constraint but in `SQLite` we extract them from whole constraints. also, the SQL expression is different in SQLite.

felixxm commented:
@hramezani Can I push my edits? I added few more tests.

felixxm commented:
@hramezani OK, I could :smile: We should split changes into two commits:

- 1st: `"Refs #30593 - Added _parse_table_constraints() hook to introspection on MariaDB"` (with a new hook and tests that work without any changes in the logic),
- 2nd: `"Refs #30593 -- Fixed introspection of check constraints columns on MariaDB."` (with fix and extra tests).

hramezani commented:
@felixxm yes, you can :smile: 
I will do the split.

felixxm commented on django/db/backends/mysql/introspection.py:
```diff
@@ -147,6 +147,16 @@ def get_storage_engine(self, cursor, table_name):
+        for token in tokens:
+            if token.ttype == sqlparse.tokens.Name:
+                if token.value[1:-1] in columns:
+                    cols.add(token.value[1:-1])
```
We assume (and it seems that it is true) that columns names are always quoted, we should use this here to fix cases when columns names overlap with function names (see tests with `LOWER`), e.g.
```python
if (
    token.ttype == sqlparse.tokens.Name and
    self.connection.ops.quote_name(token.value) == token_value and
    token.value[1:-1] in columns
):
    cols.add(token.value[1:-1])
```

hramezani replied:
Done.

felixxm commented:
@hramezani Thanks!
